librz/bin/p/bin_elf.inc

```diff
@@ -192,10 +192,117 @@ static RzBinElfNoteFile *note_file_for_load_segment(struct Elf_(rz_bin_elf_obj_t
 	return NULL;
 }
 
+static ut32 section_perms_from_flags(ut32 flags) {
+	ut32 r = 0;
+	if (RZ_BIN_ELF_SCN_IS_EXECUTABLE(flags)) {
+		r |= RZ_PERM_X;
+	}
+	if (RZ_BIN_ELF_SCN_IS_WRITABLE(flags)) {
+		r |= RZ_PERM_W;
+	}
+	if (RZ_BIN_ELF_SCN_IS_READABLE(flags)) {
+		r |= RZ_PERM_R;
+	}
+	return r;
+}
+
+static RzList *maps(RzBinFile *bf) {
+	struct Elf_(rz_bin_elf_obj_t) *obj = (bf && bf->o) ? bf->o->bin_obj : NULL;
+	RzList *ret = NULL;
+	if (!obj || !(ret = rz_list_newf((RzListFree)rz_bin_map_free))) {
+		return NULL;
+	}
+	Elf_(Phdr) *phdr = obj->phdr;
+	if (phdr) {
+		ut64 core_sp = Elf_(rz_bin_elf_get_sp_val)(obj);
+		int n = 0;
+		for (size_t i = 0; i < obj->ehdr.e_phnum; i++) {
+			if (phdr[i].p_type != PT_LOAD) {
+				continue;
+			}
+			RzBinMap *map = RZ_NEW0(RzBinMap);
+			if (!map) {
+				break;
+			}
+			map->paddr = phdr[i].p_offset;
+			map->psize = phdr[i].p_filesz;
+			map->vsize = phdr[i].p_memsz;
+			map->vaddr = phdr[i].p_vaddr;
+			map->perm = phdr[i].p_flags | RZ_PERM_R;
+			// map names specific to core files...
+			if (core_sp != UT64_MAX && core_sp >= phdr[i].p_vaddr && core_sp < phdr[i].p_vaddr + phdr[i].p_memsz) {
+				map->name = strdup("[stack]");
+			} else {
+				RzBinElfNoteFile *nf = note_file_for_load_segment(obj, &phdr[i]);
+				if (nf && nf->file) {
+					map->name = strdup(nf->file);
+				}
+			}
+			// generic names
+			if (!map->name) {
+				map->name = rz_str_newf("LOAD%d", n);
+			}
+			n++;
+			rz_list_append(ret, map);
+		}
+	} else {
+		// Load sections if there is no PHDR
+		struct rz_bin_elf_section_t *sections = Elf_(rz_bin_elf_get_sections)(obj);
+		if (sections) {
+			for (size_t i = 0; !sections[i].last; i++) {
+				RzBinMap *map = RZ_NEW0(RzBinMap);
+				if (!map) {
+					break;
+				}
+				map->name = strdup(sections[i].name);
+				map->paddr = sections[i].offset;
+				map->psize = sections[i].type != SHT_NOBITS ? sections[i].size : 0;
+				map->vsize = sections[i].size;
+				map->vaddr = sections[i].rva;
```

Comment:
Ok, I think this might be a good time to fix this.

If there is no PHDR, it means this is a relocatable file (.o) and in that case AFAIK the virtual address of the sections does not have any sense (all sections have address=0, i think). Essentially, we can place sections wherever we want, but we should not have them overlap at least. That's what the linker does in other words, as it finds a place for each section in the final file.

Reply:
@ret2libc I think it's better to do that in the separate PR since it might be a bit more code than that. We already have issue to track that: https://github.com/rizinorg/rizin/issues/815

Reply:
For reference: https://github.com/rizinorg/rizin/issues/815
I think so too. I wouldn't do this in this pr since it tries to not change behaviour unless strictly necessary, but after it is a good time. 
What we can do now with `RzBinMap` is also things like map things at a vaddr that we freely decide but still report the original 0 vaddr in the `RzBinSection` as it appears in the file. We might have to adjust some analysis/search code though in case something assumes that sections will always be mapped at the `RzBinSection` vaddr.

Reply:
Ok ok makes sense! Awesome